### PR TITLE
Implement Doctrine discriminator maps for dynamic entity management

### PR DESCRIPTION
**Pull request by james2001**
# Pull Request

## What does this PR do?
- Implements Doctrine discriminator maps to enable dynamic entity management. This enhancement addresses the issue where indexing abstract or extended entities was not possible, by allowing efficient management of different entities within the same inheritance hierarchy. This simplifies data model extension and maintenance, making the system more flexible and robust.

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Fixes the issue with indexing abstract or extended entities by utilizing Doctrine's discriminator maps, representing a significant architectural improvement.
- [x] I have read the contributing guidelines.
- [x] The title of the PR is accurate and descriptive of the changes made.

### CODE REVIEW

**Comment by norkunas** [comment]
Please add tests :)

**Comment by norkunas** [comment]
@james2001 you should rebase your PR as now your changes basically reverts https://github.com/meilisearch/meilisearch-symfony/pull/323

**Comment by james2001** [comment]
@norkunas I already rebase

image

File: src/Services/MeilisearchService.php [hunk 1]
```diff
@@ -221,6 +222,21 @@ public function shouldBeIndexed(object $entity): bool
         return true;
     }
 
+    private function getBaseClassName($object_or_class): string
```
Comment (norkunas) [inline src/Services/MeilisearchService.php:225]:
```suggestion
    /**
     * @param object|class-string $objectOrClass
     */
    private function getBaseClassName(object|string $objectOrClass): string
```

File: tests/BaseKernelTestCase.php [hunk 2]
```diff
@@ -148,7 +149,7 @@ protected function createTag(array $properties = []): Tag
 
     protected function createLink(array $properties = []): Link
     {
-        $link = new Link();
+        $link = new ExternalLink();
```
Comment (norkunas) [inline tests/BaseKernelTestCase.php:152]: It's better to create new fixtures for different features to avoid modifying existing tests

**Review by norkunas — CHANGES_REQUESTED** [review]

File: src/Services/MeilisearchService.php (hunk 1 above)
Comment (james2001) [reply src/Services/MeilisearchService.php:230]: Union type is not allowed in PHP 7.4

Comment (norkunas) [reply src/Services/MeilisearchService.php:230]: Ok so at least phpdoc and camelcased param

**Comment by norkunas** [comment]
Tests should create then at least 1 article and 1 podcast :)

**Review by norkunas — APPROVED** [review]

**Comment by norkunas** [comment]
bors merge

**Comment by norkunas** [comment]
Thank you @james2001 :)

**Pull request closed without merging**